File: tests/fixtures/v2_phase_io/typescript_express_api/delivery/input/src/routes/echo_handler.ts
// Echo Handler — 处理 POST /api/echo 端点业务逻辑

import { Request, Response, Router } from 'express';

/**
 * Echo response body structure.
 */
export interface EchoResponse {
  echo: string;
  length: number;
}

/**
 * Echo request body structure.
 */
export interface EchoRequest {
  message: string;
}

/**
 * Processes an echo request and returns the message with its character length.
 * @param message - The message string to echo back
 * @returns EchoResponse with echo field and length field
 */
export function processEcho(message: string): EchoResponse {
  return {
    echo: message,
    length: message.length,
  };
}

/**
 * Handles the POST /api/echo endpoint.
 * Validates the request body and returns the echoed message with its length.
 * Returns HTTP 400 if the request body is malformed.
 */
export function echoHandler(req: Request, res: Response): void {
  const body = req.body as EchoRequest;

  if (!body || typeof body.message !== 'string') {
    res.status(400).json({
      error: 'Invalid request body. Expected JSON with a "message" string field.',
    });
    return;
  }

  const result = processEcho(body.message);
  res.status(200).json(result);
}

/**
 * Registers the echo route on the given router.
 * @param router - Express Router instance
 */
export function registerEchoRoute(router: Router): void {
  router.post('/api/echo', echoHandler);
}

File: tests/fixtures/v2_phase_io/go_api/main_entry/input/internal/handlers/echo_handler.go
package handlers

import (
	"encoding/json"
	"net/http"
)

// EchoRequest represents the JSON body for POST /api/echo.
type EchoRequest struct {
	Message *string `json:"message"`
}

// EchoResponse represents the JSON response for POST /api/echo.
type EchoResponse struct {
	Echo   string `json:"echo"`
	Length int    `json:"length"`
}

// EchoHandler handles POST /api/echo requests.
type EchoHandler struct {
	MaxMessageLength int // Maximum allowed message length in bytes, default 10000
}

// NewEchoHandler creates a new EchoHandler with default configuration.
func NewEchoHandler() *EchoHandler {
	return &EchoHandler{
		MaxMessageLength: 10000,
	}
}

// ServeHTTP implements http.Handler for the echo endpoint.
// Accepts POST with JSON body {"message":"string"}, returns {"echo":"<message>","length":<int>}.
// Returns 400 for invalid input, 405 for wrong method.
func (h *EchoHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusMethodNotAllowed)
		json.NewEncoder(w).Encode(map[string]string{"error": "method not allowed"})
		return
	}

	var req EchoRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"error": "invalid input"})
		return
	}

	if req.Message == nil {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"error": "invalid input"})
		return
	}

	if len(*req.Message) > h.MaxMessageLength {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]string{"error": "message too long"})
		return
	}

	resp := EchoResponse{
		Echo:   *req.Message,
		Length: len(*req.Message),
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(resp)
}
